### PR TITLE
Allow for Metaculus authentication via API keys

### PR DESCRIPTION
**Pull request by uvafan**

### CODE REVIEW

**Comment by NixGD** [comment]
Have you tested what happens if a user authenticated by an API key tries to call `get_question` while filtering by `player_status`?  Since `user_id` is never set I think it would return an empty dataframe as is, but maybe it should throw an error instead.

I wonder if it's also worth including #89 in this PR as well.  I've thought about what changes are needed to make that work and would be happy to try to incorporate them into here later today if that would be helpful (or I could work on them after this is merged).

File: ergo/platforms/metaculus/metaculus.py
Comment (NixGD) [inline ergo/platforms/metaculus/metaculus.py:75]: The type hints for these should be Optional[str], right?

File: ergo/platforms/metaculus/metaculus.py
Comment (NixGD) [inline ergo/platforms/metaculus/metaculus.py:94]: What should happen when a user calls `.login()` on a Metaculus object that was set up with API authentication?

Comment (uvafan) [reply ergo/platforms/metaculus/metaculus.py:94]: Fair point. I think it would be reasonable to switch to user/pass authentication, I will implement this.

**Comment by uvafan** [comment]
Thanks for the comments @NixGD! 

> Have you tested what happens if a user authenticated by an API key tries to call `get_question` while filtering by `player_status`? Since `user_id` is never set I think it would return an empty dataframe as is, but maybe it should throw an error instead.

I've updated the PR to allow for a `user_id` to be specified when using API key authentication, and to throw an error if it hasn't been specified but a filter which needs it is specified.

> I wonder if it's also worth including #89 in this PR as well. I've thought about what changes are needed to make that work and would be happy to try to incorporate them into here later today if that would be helpful (or I could work on them after this is merged).

 I'm going to try to merge these changes ASAP as they will make the Metaculus submission flow in Elicit cleaner, but you can base a PR for fixing #89 off of this one, or wait until it's merged.

**Review by brachbach — CHANGES_REQUESTED** [review]
1. Tests are currently failing, I didn't look at why
2. I think you should add tests of the different login methods. I think for each of uname/pwd login and api key login:
    1. log in
    2. test to make sure a user-specific detail was properly set or can be retrieved. `user_id` seems like the most obvious for uname/pwd, not sure what makes the most sense for API keys
    2. check to make sure that you can access some user-specific detail from

**Comment by brachbach** [comment]
Hmm on further thought, I think the structure of this class doesn't really make sense with 2 different forms of auth

If it weren't a breaking API change, I'd propose restructuring:

1. constructor just takes `api_domain`
2. after constructing, you can call either `login_with_username_and_password` or `login_with_api_key` (you can think of better names)
3. you should also be able to do most things without logging in, we should only require login where it's actually necessary
    1. but we can leave this for a PR that addresses #89 

given that it is a breaking API change, I think we should probably just leave it as-is

**Comment by brachbach** [comment]
per Andreas on Slack, I think we should go ahead and make the change to the class structure

**Comment by uvafan** [comment]
@brachbach, I pushed the requested change to the class structure, plus error handling to reflect that API key authentication can currently only be used for prediction. I also added a test for API key auth (there are existing tests like you described for user/pass auth).

File: ergo/platforms/metaculus/metaculus.py
Comment (brachbach) [inline ergo/platforms/metaculus/metaculus.py:78]: should be `or` I think

**Review by brachbach — APPROVED** [review]
Cool, looking much better!

Reflecting on it more, `auth_method` feels overly complicated to me, and the `post` method seems unnecessarily abstract (given that it's only used once in the codebase, to `.predict`). I think instead I'd:

1. get rid of `auth_method`
2. create `is_logged_in_via_uname_pwd()` and `have_api_keys()` methods
3. get rid of the `post` method
4. In `predict`:
    1. if `is_logged_in_via_uname_pwd()`, then use the uname/password auth. If not:
    2. if `have_api_keys()`, then use the API keys. If not:
    3. error

I don't feel that this is strictly necessary to do before merging, so I'll approve. But I think I'd suggest doing it before merging.

**Comment by brachbach** [comment]
@uvafan I think this is TMI, but I also had this alternative, more abstract idea to get rid of `auth_method` and `post`:

1. don't store an `auth_method`. Instead, create `is_logged_in_via_uname_pwd()` and `have_api_keys()` methods.
2. change `post` to use whichever form of auth we have that allows us to do the most:
    1. if `is_logged_in_via_uname_pwd()`, then use the uname/password auth. If not:
    2. if `have_api_keys()`, then use the API keys. If not:
    3. try with no auth
3. if Metaculus returns `403 forbidden` for either the API keys or no auth, error to the user, "Metaculus forbids this action for the most permissive form of authentiation you currently have, which is []"

But, I think the simpler, concrete version I described in my review comment is better

Just including this in case you like it better for some reason

File: ergo/platforms/metaculus/metaculus.py
Comment (uvafan) [reply ergo/platforms/metaculus/metaculus.py:78]: Yeah I wasn't really sure which it should be, I'll just make it `or`, not sure it matters much.

**Comment by NixGD** [comment]
I pushed some tests for unauthenticated users.  Note that one currently fails (`get_questions_json` checks for `user_id`, which may not be set)

**Comment by uvafan** [comment]
@NixGD thanks for the tests! They should all pass now.

**Review by stuhlmueller — APPROVED** [review]

**Pull request merged**
